cuda/dot_kernels.cuh

```diff
@@ -0,0 +1,299 @@
+template <typename ValueType>
+void dot(myBlasHandle *handle, const matrix_info x_info, const ValueType *x,
+         const matrix_info y_info, const ValueType *y, ValueType *res)
+{
+    constexpr std::int32_t block_size{dot_block_size};
+    const dim3 block(block_size, 1, 1);
+    const dim3 grid(
+        handle->get_device_property().multiProcessorCount * grids_per_sm, 1, 1);
+
+    kernel::init_res<<<1, 1>>>(res);
+    kernel::dot<block_size, ValueType>
+        <<<grid, block>>>(static_cast<std::int32_t>(x_info.size[0]), x,
+                          static_cast<std::int32_t>(x_info.stride), y,
+                          static_cast<std::int32_t>(y_info.stride), res);
+}
+
+/**
+ * Computes the DOT product of both x and y, and stores it in res.
+ *
+ * @tparam ArType  type used for the arithmetic operations
+ * @tparam StType  type of the values both x and y are stored in
+ * @tparam ResType  type of the result value (can be different from ArType and
+ *                  StType)
+ *
+ * @param handle  RAII object for storing scheduling information and temporary
+ *                storage
+ * @param x_info  Information about the x vector
+ * @param x  x vector
+ * @param y_info  Information about the y vector
+ * @param y  y vector
+ * @param res  pointer to a single result value
+ **/
+template <typename ArType, typename StType, typename ResType>
+void acc_dot(myBlasHandle *handle, const matrix_info x_info, const StType *x,
+             const matrix_info y_info, const StType *y, ResType *res)
+{
+    constexpr std::int32_t block_size{dot_block_size};
+    const dim3 block(block_size, 1, 1);
+    const dim3 grid(
+        handle->get_device_property().multiProcessorCount * grids_per_sm, 1, 1);
+
+    // Accessor Setup
+    constexpr std::size_t dimensionality{2};
+    std::array<std::size_t, dimensionality - 1> x_stride{x_info.stride};
+    std::array<std::size_t, dimensionality - 1> y_stride{y_info.stride};
+
+    using accessor =
+        gko::acc::reduced_row_major<dimensionality, ArType, StType>;
+    using range = gko::acc::range<accessor>;
+    using c_range = gko::acc::range<typename accessor::const_accessor>;
+    auto x_acc = c_range(x_info.size, x, x_stride);
+    auto y_acc = c_range(y_info.size, y, y_stride);
+
+    ArType *acc_dot_res{nullptr};
+    constexpr bool use_temporary_storage{!std::is_same<ArType, ResType>::value};
+    if (use_temporary_storage) {
+        acc_dot_res = handle->get_device_value_ptr<ArType>();
+    } else {
+        // reinterpret_cast only necessary to make the compiler happy. They
+        // should be the same type.
+        acc_dot_res = reinterpret_cast<ArType *>(res);
+    }
+
+    kernel::init_res<<<1, 1>>>(acc_dot_res);
+    kernel::acc_dot<block_size><<<grid, block>>>(x_acc, y_acc, acc_dot_res);
+
+    // If the types differ, we need a cast in the end to write it in the correct
+    // type
+    if (use_temporary_storage) {
+        kernel::cast_result<<<1, 1>>>(acc_dot_res, res);
+    }
+}
```

Comment:
Yes, exactly. Since I hide the temporary storage behind the `myBlasHandle`, I can ensure that the atomic add can always be performed using `ar_type`, and when `st_type` is something different, I will do a final cast at the very end, so the output is `st_type`.